### PR TITLE
network/metrics: Expose number of banned peers from peerstore and enable litep2p metrics

### PR DESCRIPTION
**Pull request by lexnv**
This PR extends the metrics exposed by the peerstore with the total number of banned peers.

The new metric is exposed under `substrate_sub_libp2p_peerset_num_banned_peers`.

To easily extend metrics in the future, the `fn num_known_peers` is removed in favor of `fn status`.

While at it, enable the metrics for litep2p:
- total number of peers from peerstore (needed to debug memory consumption)
- total number of banned peers from peerstore (needed to debug reputation bans and disconnects)

Have added a couple of tests to validate that the number of banned peers is exposed properly.

Part of: https://github.com/paritytech/polkadot-sdk/issues/4681

### Testing Done
Using [subp2p-explorer](https://github.com/lexnv/subp2p-explorer) have submitted random data on tx protocol.
The peer gets banned, the num of banned peers is incremented then the peer is disconnected.

cc @paritytech/networking

### CODE REVIEW

File: substrate/client/network/src/peer_store.rs [hunk 1]
```diff
@@ -95,14 +107,11 @@ pub trait PeerStoreProvider: Debug + Send + Sync {
...
-	///
-	/// This number might not include some connected peers in rare cases when their reputation
-	/// was not updated for one hour, because their entries in [`PeerStore`] were dropped.
-	fn num_known_peers(&self) -> usize;
-
 	/// Add known peer.
 	fn add_known_peer(&self, peer_id: sc_network_types::PeerId);
+
+	/// Get the status of the peer store.
+	fn status(&self) -> PeerStoreStatus;
```
Comment (bkchr) [inline substrate/client/network/src/peer_store.rs:114]: Looking at the implementation of this function. Why don't we update the metrics directly in the peerstore? Then we don't need to lock any mutexes?

File: substrate/client/network/src/litep2p/peerstore.rs [hunk 2]
```diff
@@ -270,21 +274,25 @@ pub struct Peerstore {
...
-	/// Create new [`Peerstore`] from a [`PeerstoreHandle`].
-	pub fn from_handle(peerstore_handle: PeerstoreHandle, bootnodes: Vec<PeerId>) -> Self {
-		for bootnode in bootnodes {
-			peerstore_handle.add_known_peer(bootnode);
-		}
+	pub fn new(bootnodes: Vec<PeerId>, metrics_registry: Option<Registry>) -> Self {
+		let metrics = if let Some(registry) = &metrics_registry {
+			PeerSetMetrics::register(registry)
+				.map_err(|err| {
+					log::error!(target: LOG_TARGET, "Failed to register peer set metrics: {}", err);
```
Comment (dmitry-markin) [inline substrate/client/network/src/litep2p/peerstore.rs:281]:
```suggestion
					log::error!(target: LOG_TARGET, "Failed to register peer store metrics: {}", err);
```
(`Peer set` is per protocol.)

File: substrate/client/network/src/litep2p/peerstore.rs [hunk 3]
```diff
@@ -270,21 +274,25 @@ pub struct Peerstore {
...
+					err
+				})
+				.ok()
+		} else {
+			None
+		};
+
+		let peerstore_inner = PeerstoreHandleInner {
```
Comment (dmitry-markin) [inline substrate/client/network/src/litep2p/peerstore.rs:289]: nit: May be add/use a constructor for `PeerstoreHandle` and not initialize `PeerstoreHandleInner` manually?

File: substrate/client/network/src/service/metrics.rs [hunk 4]
```diff
@@ -225,6 +220,34 @@ impl Metrics {
 	}
 }
 
+/// PeerSet metrics.
+#[derive(Clone, Debug)]
+pub struct PeerSetMetrics {
```
Comment (dmitry-markin) [inline substrate/client/network/src/service/metrics.rs:225]: I would say it is `PeerStoreMetrics`, as it is common to all protocols. Currently we use "peer set" to refer to specific notifications protocol's peer set.

File: substrate/client/network/src/protocol/notifications/tests.rs [hunk 5]
```diff
@@ -76,11 +76,14 @@ fn build_nodes() -> (Swarm<CustomProtoWithAddr>, Swarm<CustomProtoWithAddr>) {
...
 		let (protocol_handle_pair, mut notif_service) =
 			crate::protocol::notifications::service::notification_service("/foo".into());
-		let peer_store = PeerStore::new(if index == 0 {
-			keypairs.iter().skip(1).map(|keypair| keypair.public().to_peer_id()).collect()
-		} else {
-			vec![]
-		});
+		let peer_store = PeerStore::new(
+			if index == 0 {
+				keypairs.iter().skip(1).map(|keypair| keypair.public().to_peer_id()).collect()
```
Comment (dmitry-markin) [inline substrate/client/network/src/protocol/notifications/tests.rs:81]: This seems cryptic. While you are at it, may be add a comment?

File: substrate/client/network/src/service.rs [hunk 6]
```diff
@@ -1020,7 +1023,7 @@ where
 	}
 
 	fn report_peer(&self, peer_id: sc_network_types::PeerId, cost_benefit: ReputationChange) {
-		self.peer_store_handle.clone().report_peer(peer_id, cost_benefit);
+		self.peer_store_handle.report_peer(peer_id, cost_benefit);
```
Comment (dmitry-markin) [inline substrate/client/network/src/service.rs:1026]: I wonder what was the purpose of that `clone()` and why it's not needed anymore?

**Review by dmitry-markin — APPROVED** [review]

File: substrate/client/network/src/service.rs (hunk 6 above)
Comment (lexnv) [reply substrate/client/network/src/service.rs:1026]: The peer store handle is behind an Arc `peer_store_handle: Arc<dyn PeerStoreProvider>`, therefore the cloning should be cheap here.  I think this was a miss, there's no need to clone the handler to access its methods (we do have other calls to report_peer which don't clone)

File: substrate/client/network/src/litep2p/peerstore.rs (hunk 3 above)
Comment (lexnv) [reply substrate/client/network/src/litep2p/peerstore.rs:289]: Yep that makes sense, thanks!

File: substrate/client/network/src/service/metrics.rs (hunk 4 above)
Comment (lexnv) [reply substrate/client/network/src/service/metrics.rs:225]: Will rename thanks! 🙏

**Review by alexggh — APPROVED** [review]
Looks good to me!

**Review by bkchr — APPROVED** [review]

**Comment by kianenigma** [comment]
I am going to repeat exactly what I said in https://github.com/paritytech/polkadot-sdk/pull/2944#issuecomment-2112060450 + complain about the lack of:  

https://github.com/paritytech/polkadot-sdk/blob/master/docs/contributor/PULL_REQUEST_TEMPLATE.md#integration

This PR should have not been marked as Silent. It _is_ a breaking change in many ways :)

**Pull request merged**